perceval/components/port.py

```diff
@@ -158,31 +190,11 @@ def is_connected_to(self, component) -> bool:
         return component in self._connections
 
 
-class LogicalState(list):
-    def __init__(self, state: List[int]):
-        assert state.count(0) + state.count(1) == len(state), "A logical state should only contain 0s and 1s"
-        super().__init__(state)
-
-    def to_basic_state(self, port_list: List[APort]) -> BasicState:
-        result = []
-        index = 0
-        for port in port_list:
-            if isinstance(port, Herald):
-                continue
-            if index >= len(self):
-                raise ValueError('Logical state and port list do not match (state too short)')
-            if isinstance(port, Port):
-                if port.encoding == Encoding.RAW or port.encoding == Encoding.TIME:
-                    result += [self[index]]
-                    index += 1
-                elif port.encoding == Encoding.DUAL_RAIL:
-                    result += [1, 0] if self[index] == 0 else [0, 1]
-                    index += 1
-                elif port.encoding == Encoding.QUDIT or port.encoding == Encoding.POLARIZATION:
-                    raise NotImplementedError
-        if index != len(self):
-            raise ValueError('Logical state and port list do not match (state too long)')
-        return BasicState(result)
-
-    def __str__(self):
-        return '|' + ','.join([str(x) for x in self]) + '>_L'
+def get_BS_from_ports(ports : list[APort], state: LogicalState) -> BasicState:
+    logical_state = LogicalState()
+    port_list = [port for port in ports if port.has_logical_state_equivalent()]
+    if len(port_list) != len(state):
+        raise ValueError('Logical state and port list size do not match')
+    for port, mode in zip(port_list, state):
+        logical_state += port.to_logic_state(mode)
+    return BasicState(logical_state)
```

Comment:
I think there's a misunderstanding. This line should not happen, because it's semantically wrong. Why "casting" a logical state to a basic state: they represent different things and are not equal for all encodings except one (RAW).